### ISSUE TITLE
CA1062 False positive because [NotNullWhen] is not respected

### ISSUE THREAD

**Issue by RobSiklos** [dotnet/roslyn-analyzers#6990]
### Analyzer

**Diagnostic ID**: [CA1062](https://learn.microsoft.com/en-ca/dotnet/fundamentals/code-analysis/quality-rules/ca1062): `Validate arguments of public methods`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 5 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: Visual Studio 17.7.0. Project using .NET 6.0 runtime.

### Describe the bug

CA1062 complains that a parameter needs to be validated as non-null before using it, despite the fact that it has been validated already via a method which makes use of [NotNullWhenAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=net-7.0)

### Steps To Reproduce

1. Create a new C# class library project using .NET 6.0
2. Enable the CA1062 rule. (e.g. add the following to `.editorconfig`:
`dotnet_diagnostic.CA1062.severity = warning`
3. Add the following code:
```csharp
#nullable enable

using System.Diagnostics.CodeAnalysis;

namespace NotNullTest1
{
    internal static class Util
    {
        public static bool HasValue<T>([NotNullWhen(true)] T? obj)
            where T : class
        {
            return (obj != null);
        }
    }

    public class Class1
    {
        public void Foo(string s)
        {
            bool _ = Util.HasValue(s)
                     &&
                     s.Length > 5; // CA1062 triggered here
        }
    }
}
```

### Expected behavior
No warnings

### Actual behavior
CA1062 triggered on the line where `s` is referenced.

## Additional context

The warning should not be triggered, because `HasValue()` explicitly says that if the method returns `true`, it means that the argument (`s`) is not null.
